### PR TITLE
Limit regalloc prefer weight to 1.

### PR DESCRIPTION
**Pull request by ericpts**
Currently, the graph-coloring preference algorithm starts out with a default value of 8, and then each time it recurses into an if's branch, the weight is divided by 2.

Therefore, code which is nested under 4 if's will have weight 0, and no more preference information will be recorded -- even if under these if's there are maybe for loops, the `8 * weight` will still be 0.

One example of how this causes problems:

This is how the intermediary representation looks like
```
V/1 := 1
%[RDI] := V/1
call caml_sys_time_unboxed() # where %RDI is the parameter
```
If this block of code is not too deep, then it generates

```asm
	movq	$1, %rdi
	call	caml_sys_time_unboxed@PLT
```

However, if it is nested within if's, (thus triggering weight to become 0), then it generates

```asm
	movq	$1, %rax
	movq	%rax, %rdi
	call	caml_sys_time_unboxed@PLT
```
since `%rax` is the default register to allocate.

This particular problem is quite widespread across Jane Street's codebase, as it appears in more than 100k different spots.

### CODE REVIEW

**Comment by alainfrisch** [comment]
The change looks reasonable.

- Have you been able to observe actual significant speed and/or code size gains with the change?

- Is it a good idea to reduce the "weight" under branches?  Why should nested if-then-else use different weights than, say, an n-ary switch?  It would be good to see real-world impact of just dropping the `/ 2` altogether.

File: asmcomp/interf.ml [hunk 1]
```diff
@@ -148,6 +148,7 @@ let build_graph fundecl =
   (* Compute preferences and spill costs *)
 
   let rec prefer weight i =
+    let weight = max 1 weight in
```
Comment (alainfrisch) [inline asmcomp/interf.ml]: This adds some "floor" logic to the weight.  There is a similar "cap", which is enforced on each call site that bumps the weight (`* 8`).  It would make sense to be coherent between how the cap and the floor are implemented: either do both when entering `prefer`, or enforce them "manually" on each call site that changes the weight.

**Comment by alainfrisch** [comment]
Also, if the notion of weight is to be kept, I wonder (1) whether it makes sense to keep the weight unchanged for the `Itrywith` handler; (2) same for `Icatch` which is just another kind of switch.

**Comment by dra27** [comment]
cc @ctk21

File: asmcomp/interf.ml (hunk 1 above)
Comment (ericpts) [reply asmcomp/interf.ml]: Solved, by also upper capping it at 1000 in the entry.

**Comment by ericpts** [comment]
I have tested the latest changes (including not decreasing the weight at all in if's), and it seems that it had decreased code size; I have attached a zip file containing the results for some benchmarks we ran -- these were run with flambda.
[interf.zip](https://github.com/ocaml/ocaml/files/3569112/interf.zip)

`ceil` is only with lower bounding weight at 1.
`if` is with keeping the weight the same when recursing into an if's body.

The most significant stat to look at is `size_text`, where there is a constant decrease of code size of about 0.2% throughout.

There is also a gain in speed in one particular benchmark, `valet`, of something like 10%. I suspect that this is mostly noise though, and not representative.


Regarding also not decaying the weighs inside a `switch`'s body, results for that should come soon.

**Comment by ericpts** [comment]
Update with new results
[interf-2.html.gz](https://github.com/ocaml/ocaml/files/3595619/interf-2.html.gz);
`switch` contains keeping the same weight with if's, and also with switches.

There is a consistent reduction of size, of roughly 0.3%; and also speed increases.

The first set of benchmarks were done with cross-library inlining accidentally turned off, so these latest results should be representative of the actual gains.

**Comment by ctk21** [comment]
I've taken a look at the interf-2.html.gz results. 

If you take `perf_instructions` as the metric and hit the normalize button, it seems like the `base` variant has more than 12 times the number of executed instructions for `qr-decomposition`. 

Assuming I've interpreted the results correctly, do you know what in this change is causing such a dramatic reduction in instructions executed for `qr-decomposition` when you are seeing ~0.3% code size reduction more generally?

**Comment by ericpts** [comment]
Yes you are correct, there was a mistake when computing them -- the base compiler had some extra instrumentation leftover from previous benchmarks.

This version should be correct:
[interf-3.html.gz](https://github.com/ocaml/ocaml/files/3599962/interf-3.html.gz)

**Comment by xavierleroy** [comment]
Some comments (better late than never):
* Agreed, preferences should not vanish when their weight drops below 1.  That was a bug, thanks for the fix.
* The idea of putting weights on preference edges comes from Preston Briggs' PhD work, which was the state of the art in coloring-based register allocation at the time ocamlopt's register allocator was designed.  Briggs proposed to put bigger weights on moves contained within loops, on the basis that they are executed more frequently.  Putting lower weights on moves from "then" and "else" branches of conditionals is the same logic: often these moves are executed less often than moves outside of the conditional.
* The logic above is probably flawed to begin with.  Coalescing moves is not just about reducing the run-time costs of moves (which are low), but also reducing the code size.  Plus, we don't have a good static estimation of how frequently a move is executed.
* As another case in point, the IRC algorithm of George and Appel does not put weights on preference edges, treating them all with equal priority.

In summary:
* I agree with not decreasing weights for if/then/else.  Only loops increase weights, like in Briggs' approach.
* Later, we could consider removing weights altogether.  It would simplify the register allocator and could decrease compilation times with no impact on code quality.

File: asmcomp/interf.ml (hunk 1 above)
Comment (xavierleroy) [reply asmcomp/interf.ml]: Now that weight is never divided by two, do we still need this "max 1" safeguard?

**Review by xavierleroy — COMMENTED** [review]
I think this is good to go, except that the "weight is at least 1" logic is now useless, or so it seems.

**Comment by xavierleroy** [comment]
With apologies to @alainfrisch, I wasn't convinced by the "cap weight to 1000 at function entry" logic.  For one thing, the `min` operation is not cheap (because of polymorphic comparison), while the weight changes only at one recursive call.  

I took the liberty of pushing a commit that reverts this particular change, then simplifies further the computation of the new weight to be used for recursive handlers.  

I'll do a run of CI precheck.  If it's good and nobody complains, I'll merge.

**Comment by alainfrisch** [comment]
> With apologies to @alainfrisch, I wasn't convinced by the "cap weight to 1000 at function entry" logic. For one thing, the `min` operation is not cheap (because of polymorphic comparison), while the weight changes only at one recursive call.

I just wanted to avoid having a different treatment between the cap and the floor (but the floor is gone, so my remark is void).

But the fact that min/max of integers go through the polymorphic comparison is concerning in itself (and confirmed empirically).  Do you know why `Stdlib.min` is not inlined?  (This should make it possible to re-specialize the comparison primitive in Cmm, at least when one of the arguments is a constant.)

**Comment by xavierleroy** [comment]
Precheck CI is green, Travis and AppVeyor are happy: stars are aligned, let's merge.

**Comment by xavierleroy** [comment]
> Do you know why Stdlib.min is not inlined?

It is inlined with the flambda back-end and high enough `-O` option, but the inlined code still contains a call to the generate "greater than or equal" primitive, so it's not going to run any faster than the code without inlining.

With the default back-end, I think the definition of Stdlib.min falls just above the inlining threshold. 

> (This should make it possible to re-specialize the comparison primitive in Cmm, at least when one of the arguments is a constant.)

I think users understand (and expect) type-driven specialization, not constant-driven specialization.

**Comment by alainfrisch** [comment]
> I think users understand (and expect) type-driven specialization, not constant-driven specialization.

Right, but `fun x -> x = None` is already being optimized, which looks like constant-driven optimization of the comparison primitive.
